### PR TITLE
Feature higher spatial order

### PR DESCRIPTION
**Pull request by tj-sun**
Higher spatial order

- use -so in command line to generate code for different spatial order (default to 2). e.g. -so 4 to generate 8th order
- fixed previous bug on AI calculation
- simple algorithm for free-surface boundary for now (setting velocity to zero), according to Robertsson's 1996 paper "A numerical free-surface condition for elastic/viscoelastic finite-difference modeling in the presence of topography"
- helper functions to compute non-symmetric FD approximation, which is needed for other algorithm for free surface

### CODE REVIEW

File: opesci/fields.py [hunk 1]
```diff
@@ -201,51 +202,70 @@ def set(self, dimension, direction):
...
-        rhs = solve(eq, lhs)[0]
-        lhs = lhs.subs(self.indices[d], t)
-        rhs = self.align(rhs.subs(self.indices[d], t))
-
-        # change ti to t+1
-        lhs = lhs.subs(idx[0], idx[0]+1)
-        rhs = rhs.subs(idx[0], idx[0]+1)
-
-        self.bc[d][side] = [Eq(lhs, rhs)]
+        if algo == 0:
```
Comment (ctjacobs) [inline opesci/fields.py]: Document what each algorithm is supposed to do under each branch of the if-statement, i.e.

```
if algo == 0:
   # This does foo.
```

Comment (ctjacobs) [inline opesci/fields.py]: Also, can we use string names for the algorithms rather than some magic numbers.

File: opesci/fields.py [hunk 2]
```diff
@@ -201,51 +202,70 @@ def set(self, dimension, direction):
...
+                idx[d] = b - (-1)**side
+            eq = Eq(self[idx])
+            eq = eq.subs(idx[0], idx[0]+1)
+            self.bc[d][side] = [eq]
+            # populate all ghost cells
+            for depth in range(self.accuracy[d]-1):
+                idx[d] -= (-1)**side
+                eq = Eq(self[idx])
+                eq = eq.subs(idx[0], idx[0]+1)
+                self.bc[d][side].append(eq)
```
Comment (ctjacobs) [inline opesci/fields.py]:
```
else:
   raise ValueError("Unknown ghost cell algorithm")
```

File: opesci/fields.py [hunk 3]
```diff
@@ -291,8 +311,9 @@ def set_free_surface(self, d, b, side):
         store the code to populate ghost cells to self.bc
         """
```
Comment (ctjacobs) [inline opesci/fields.py]: Also document what `algo` does here.

File: opesci/util.py [hunk 4]
```diff
@@ -137,65 +152,84 @@ def Deriv(U, i, k, d, n):
...
-    :param i: list of indices (Symbol objects) of field U
+    :param index: list of indices (Symbol objects) of field U
     e.g. possibley [t,x,y,z] for 3D
     :param k: determine the derivative is with respect to which dimension
     e.g. dU/dt if k=0, dU/dx if k=0, assuming i=[t,x,y,z]
-    :param d: spacing of the grid, e.g. dx, dt
-    :param n: accuracy of approximation
-    e.g. n=1 for 2nd-order FD approximation, n=2 for 4th-order FD approximation
+    :param delta: spacing of the grid, e.g. dx, dt
+    :param order: order of accuracy of approximation/2
```
Comment (ctjacobs) [inline opesci/util.py]: This is really confusing. By `order` I assume you mean the order of the derivative?

File: tests/eigenwave3d.py [hunk 5]
```diff
@@ -246,6 +252,8 @@ def main():
                        formatter_class=RawTextHelpFormatter)
     p.add_argument('mode', choices=('default', 'read', 'converge', 'cx1'),
                    nargs='?', default='default', help=ModeHelp)
+    p.add_argument('-so', '--spatial_order', default=2, type=int, dest='so',
+                   help='Accuracy order to use for code generation, eg. 2 for 4th order in x,y,z')
```
Comment (ctjacobs) [inline tests/eigenwave3d.py]: "Accuracy order" is confusing. From the name of the argument, you mean the order of the spatial discretisation. The fourth order convergence that you see is the (expected) order of the solution's convergence. You should make this distinction.

File: opesci/util.py [hunk 6]
```diff
@@ -118,13 +118,28 @@ def Taylor_half(dx, n):
...
+
+
+def Deriv(U, index, k, d, n):
     """
     calculate the FD approximation for nth derivative
     the function works by using M * D = R, as specified in Taylor()
     then inverting M, so that D = M_inverse * R
     :param U: the field
-    :param i: list of indices (Symbol objects) of field U
+    :param index: list of indices (Symbol objects) of field U
```
Comment (ctjacobs) [inline opesci/util.py]: Note that you can also specify the data type in between `param` and the `param`'s name, e.g. `:param list index:`

File: opesci/staggeredgrid.py [hunk 7]
```diff
@@ -435,12 +436,16 @@ def set_free_surface_boundary(self, dimension, side):
...
         """
+        algo = 0
+        if self.order[dimension] > 2:
+            # using different algorithm for free surface for higher order (>4)
+            algo = 1
         self.associate_fields()
         for field in self.sfields+self.vfields:
             if side == 0:
-                field.set_free_surface(dimension, self.margin.value, side)
+                field.set_free_surface(dimension, self.margin.value, side, algo)
```
Comment (ctjacobs) [inline opesci/staggeredgrid.py]: Being picky: `algo=algo`

File: opesci/staggeredgrid.py [hunk 8]
```diff
@@ -435,12 +436,16 @@ def set_free_surface_boundary(self, dimension, side):
...
             else:
-                field.set_free_surface(dimension, self.dim[dimension-1]-self.margin.value-1, side)
+                field.set_free_surface(dimension, self.dim[dimension-1]-self.margin.value-1, side, algo)
```
Comment (ctjacobs) [inline opesci/staggeredgrid.py]: Being picky: `algo=algo`

**Comment by mlange05** [comment]
Can we please add a few higher-order runs to our test setup and check manually that we achieve the convergence we would expect given the temporal and spatial discretisation?

File: opesci/fields.py (hunk 1 above)
Comment (tj-sun) [inline opesci/fields.py]: Added names and references

File: opesci/fields.py (hunk 2 above)
Comment (tj-sun) [inline opesci/fields.py]: done

File: opesci/fields.py (hunk 3 above)
Comment (tj-sun) [inline opesci/fields.py]: done

File: opesci/util.py (hunk 4 above)
Comment (tj-sun) [inline opesci/util.py]: It actually means order of FD approximation (see example in next line). The function returns a list of FD approximation of different orders of derivatives (to the same accuracy order)

File: tests/eigenwave3d.py (hunk 5 above)
Comment (tj-sun) [inline tests/eigenwave3d.py]: done

File: opesci/util.py (hunk 6 above)
Comment (tj-sun) [inline opesci/util.py]: added.
these functions hopefully will be replaced by the generic versions later.

File: opesci/staggeredgrid.py (hunk 7 above)
Comment (tj-sun) [inline opesci/staggeredgrid.py]: amended

File: opesci/staggeredgrid.py (hunk 8 above)
Comment (tj-sun) [inline opesci/staggeredgrid.py]: amended

**Comment by tj-sun** [comment]
comparing convergence between different spatial order is quite hard to do currently, as different algorithms are used for free-surfaces. The errors for algorithm used for >4th order are quite big currently. This need to be fixed in another branch.

**Comment by ggorman** [comment]
"use -so in command line to generate code for different spatial order (default to 2). e.g. -so 4 to generate 8th order"

I've raised this a few times (see #41). It might make sense to you in terms of the implementation to say "set 4 if you mean 8" -- but not to a someone developing a model or using who would expect "set 8 if you mean 8". You can work out the logic within the code - so please allow the interface and API to set the order as would be commonly understood.

RE: Principle of least astonishment ;-)

**Comment by tj-sun** [comment]
Re-factored set_accuracy()

**Comment by mlange05** [comment]
This looks good and is long overdue, so merging...

**Pull request merged**